src/html-imports.js

```diff
@@ -7,720 +7,756 @@
+          n.textContent += `\n//# sourceURL=${url}${num}.js\n`;
+          inlineScriptIndex++;
         }
       }
+      Path.fixUrlsInTemplates(content, url);
+      return content;
+    }
+
+    onLoadedAll() {
+      this.flatten(document);
+      // We wait for styles to load, and at the same time we execute the scripts,
+      // then fire the load/error events for imports to have faster whenReady
+      // callback execution.
+      // NOTE: This is different for native behavior where scripts would be
+      // executed after the styles before them are loaded.
+      // To achieve that, we could select pending styles and scripts in the
+      // document and execute them sequentially in their dom order.
+      Promise.all([this.waitForStyles(), this.runScripts()])
+        .then(() => this.fireEvents());
     }
-  },
-
-// TODO(valdrin): if possible, fire in same order as native imports which is
-// probably bottom up.
-  _fireEvents: function(element) {
-    var n$ = element.querySelectorAll(IMPORT_SELECTOR);
-    for (var i=0; i < n$.length; i++) {
-      var n = n$[i];
-      if (!n.__loaded) {
-        flags.log && console.warn('fire', n.import ? 'load' : 'error', n.href);
-        n.__loaded = true;
-        n.dispatchEvent(new CustomEvent(n.import ? 'load' : 'error'));
+
+    /**
+     * @param {!HTMLDocument} doc
+     */
+    flatten(doc) {
+      const n$ = /** @type {!NodeList<!HTMLLinkElement>} */
+        (doc.querySelectorAll(importSelector));
+      for (let i = 0, l = n$.length, n; i < l && (n = n$[i]); i++) {
+        const imp = this.documents[n.href];
+        n.import = /** @type {!Document} */ (imp);
+        if (imp && imp.nodeType === Node.DOCUMENT_FRAGMENT_NODE) {
+          // We set the .import to be the link itself, and update its readyState.
+          // Other links with the same href will point to this link.
+          this.documents[n.href] = n;
+          n.readyState = 'loading';
+          // Suppress Closure warning about incompatible subtype assignment.
+          ( /** @type {!HTMLElement} */ (n).import = n);
+          this.flatten(imp);
+          n.appendChild(imp);
+        }
       }
     }
-  },
-
-  observe: function(element) {
-    if (!element.__importObserver) {
-      element.__importObserver = new MutationObserver(function (mxns) {
-        mxns.forEach(function(m) {
-          if (m.addedNodes) {
-            for (var i=0; i < m.addedNodes.length; i++) {
-              var p = m.addedNodes[i];
-              // TODO(sorvell): x-platform matches
-              if (p.nodeType === Node.ELEMENT_NODE &&
-                  p.matches(IMPORT_SELECTOR)) {
-                importer.loadNode(p);
-              }
-            }
+
+    /**
+     * Replaces all the imported scripts with a clone in order to execute them.
+     * Updates the `currentScript`.
+     * @return {Promise} Resolved when scripts are loaded.
+     */
+    runScripts() {
```

Comment:
Re-ping on status of debugging when scripts are run via this method. I believe @kevinpschaaf asked about this.